File: server/common/connection.py
import logging
def send(connection, message):
    """
    Send a message without short-writes with a \0 at the end
    """
    message = message + "\0"
    
    connection.sendall(message.encode("utf-8"))


def read_up_to_delimiter(connection, delimiter):
    """
    Read up to the delimiter, this function doesn't have short-reads
    """
    buffer = bytearray()
    found_delimiter = False
    delimiter = delimiter.encode("utf-8")
    delimiter_index = 0

    while not found_delimiter:
        chunk = connection.recv(2)
        if not chunk:
            raise ConnectionResetError("Socket is closed in reading")
        buffer.extend(chunk)
        logging.info(f"action: SERVER READING | result: success | message {buffer.decode("utf-8")}")

        delimiter_index = buffer.find(delimiter)
        if delimiter_index != -1:
            found_delimiter = True

    return buffer[:delimiter_index].decode("utf-8")

File: client/common/connection.go
package common

import (
	"bytes"
	"io"
	"net"
)

// Read up to the delimiter, this function doesn't have short-reads
// Returns a Message object with the logic depending of the message received
func readUpToDelimiter(connection net.Conn, delimiter string) (string, error) {

	buffer := make([]byte, 1024)
	var result bytes.Buffer
	foundDelimiter := false
	delimiterBytes := []byte(delimiter)
	delimiterIndex := 0

	for !foundDelimiter {
		bytesRead, err := connection.Read(buffer)
		if err != nil {
			if err == io.EOF {
				break
			}
			return "", err
		}

		result.Write(buffer[:bytesRead])

		delimiterIndex = bytes.Index(result.Bytes(), delimiterBytes)
		if delimiterIndex != -1 {
			foundDelimiter = true

		}
	}

	return string(result.Bytes()[:delimiterIndex]), nil
}

// Send a message without short-writes with a \0 at the end
func send(connection net.Conn, messageToSend string) error {
	messageToSend += "\000"
	bytesToWrite := len(messageToSend)
	bytesAlreadyWritten := 0

	for bytesAlreadyWritten < bytesToWrite {
		bytesWritten, err := connection.Write([]byte(messageToSend[bytesAlreadyWritten:]))
		if err != nil {
			return err
		}

		bytesAlreadyWritten += bytesWritten
	}

	return nil
}
